main/src/ca/uwaterloo/flix/api/lsp/provider/completion/ModuleCompleter.scala

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2023 Lukas Rønn
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp.provider.completion
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.api.lsp.Index
+import ca.uwaterloo.flix.language.ast.TypedAst
+import ca.uwaterloo.flix.api.lsp.provider.completion.Completion.ModCompletion
+import ca.uwaterloo.flix.language.ast.Symbol.ModuleSym
+
+object ModuleCompleter extends Completer {
+  /**
+    * Returns a List of Completion for completer.
+    */
+  override def getCompletions(context: CompletionContext)(implicit flix: Flix, index: Index, root: TypedAst.Root, delta: DeltaContext): Iterable[ModCompletion] = {
+    // We don't know if the user has provided a fqn, or is currently writing it so we have to try both cases
+    val word = context.word.split('.').toList
+
+    getCurrentModuleCompletion(word) ++
+      getNextModuleCompletion(word)
+  }
+
+  /**
+    * Gets completions for modules when not having a fqn
+    */
+  private def getCurrentModuleCompletion(word: List[String])(implicit root: TypedAst.Root): Iterable[ModCompletion] = {
+    // The user has provided a subWord
+    // We know that there is at least one dot, so we split the word and
+    // make a fqn from the everything but the the last (hence it is the subWord).
+    val fqnWithOutLastWord = word.dropRight(1)
+    val subWord = word.takeRight(1)(0)
```

Comment:
This is unsafe right, could give out of bounds.

Reply:
It's not unsafe, it would just produce an empty String.

Reply:
That's why i'm not using `.head`, which is unsafe.

Reply:
What if takeright returns the empty string?

Reply:
If `takeRight(1)(0)` is the empty string?

Reply:
What is the type of `takeRight(1)`?